switchy/apps/routers.py

```diff
@@ -40,3 +56,50 @@ def on_park(self, sess):
     def on_bridge(self, sess):
         self.log.debug("Bridged aleg session '{}' to bleg session '{}'"
                        .format(sess.uuid, sess['Bridge-B-Unique-ID']))
+
+
+@app
+class Router(object):
+    '''Route sessions using registered callback functions (decorated as
+    "routes") which are pattern matched based on selected channel variable
+    contents.
+
+    Requires that the handling SIP profile had been configured to use the
+    'switchy' dialplan context or at the very least a context which contains a
+    park action extension.
+    '''
+    # default routes
+    route = utils.PatternCaller()
+
+    def __init__(self, guards, use_defaults=True):
+        self.guards = guards or {}
+        self.route = utils.PatternCaller()
+        if use_defaults:
+            self.route.update(type(self).route)
+
+    def prepost(self, pool):
+        self.host = pool.evals('client.host')
+        self.log = utils.get_logger(utils.pstr(self))
+
+    @event_callback("CHANNEL_PARK")
+    def on_park(self, sess):
+        if all(sess[key] == val for key, val in self.guards.items()):
+            self.route.call_matches(sess, sess=sess, router=self)
+        else:
+            self.log.warn("Session with id {} did not pass guards"
+                          .format(sess.uuid))
+            sess.hangup('CALL_REJECTED')
```

Comment:
404 seems more suitable when there's no routes (I guess that'd be CALL_NO_ROUTE_DESTINATION (looking at the C enum, not sure where the str to int translation is done in this case from ESL)